File: go-backend-interview-prep/03_functions_and_methods/01_closures/main.py
# Python equivalent of Closures (compare with main.go)

def make_adder(x: int):
    """Returns a closure that adds x to its argument."""
    return lambda y: x + y


def counter_gen():
    """Returns a closure that counts up from 1."""
    n = 0
    def next_val():
        nonlocal n  # required to modify outer variable in Python
        n += 1
        return n
    return next_val


def main():
    # --- Basic closure: captures outer variable ---
    # Python requires 'nonlocal' to modify (Go does not)
    counter = 0
    def increment():
        nonlocal counter
        counter += 1
        return counter

    print("call 1:", increment())
    print("call 2:", increment())
    print("call 3:", increment())
    print("counter:", counter)

    # --- Closure as return value (function factory) ---
    print("\n--- Function factory ---")
    add_five = make_adder(5)
    add_ten = make_adder(10)
    print("add_five(3):", add_five(3))
    print("add_ten(3):", add_ten(3))

    # --- Closure with state ---
    print("\n--- Stateful closure ---")
    next_val = counter_gen()
    print(next_val())
    print(next_val())
    print(next_val())

    # --- Loop variable trap (same issue in Python!) ---
    print("\n--- Loop variable trap ---")
    funcs = []
    for i in range(5):
        funcs.append(lambda: print(i, end=" "))  # captures i by reference

    print("trap: ", end="")
    for fn in funcs:
        fn()  # all print 4
    print()

    # --- Fix: default argument captures current value ---
    funcs2 = []
    for i in range(5):
        funcs2.append(lambda i=i: print(i, end=" "))  # default arg = copy

    print("fixed:", end=" ")
    for fn in funcs2:
        fn()
    print()


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/03_functions_and_methods/01_closures/main.go
package main

import "fmt"

func main() {
	// --- Basic closure: captures outer variable ---
	counter := 0
	increment := func() int {
		counter++ // modifies the outer variable
		return counter
	}
	fmt.Println("call 1:", increment()) // 1
	fmt.Println("call 2:", increment()) // 2
	fmt.Println("call 3:", increment()) // 3
	fmt.Println("counter:", counter)    // 3

	// --- Closure as return value (function factory) ---
	fmt.Println("\n--- Function factory ---")
	addFive := makeAdder(5)
	addTen := makeAdder(10)
	fmt.Println("addFive(3):", addFive(3))   // 8
	fmt.Println("addTen(3):", addTen(3))     // 13

	// --- Closure with state ---
	fmt.Println("\n--- Stateful closure ---")
	next := counter_gen()
	fmt.Println(next()) // 1
	fmt.Println(next()) // 2
	fmt.Println(next()) // 3

	// --- Loop variable trap (pre-Go 1.22) ---
	fmt.Println("\n--- Loop variable trap ---")
	funcs := make([]func(), 5)
	for i := 0; i < 5; i++ {
		funcs[i] = func() {
			fmt.Print(i, " ") // captures i by reference
		}
	}
	fmt.Print("trap (Go 1.22+ fixes this): ")
	for _, fn := range funcs {
		fn()
	}
	fmt.Println()

	// --- Fix: shadow the variable ---
	fmt.Print("fixed: ")
	funcs2 := make([]func(), 5)
	for i := 0; i < 5; i++ {
		i := i // shadow: creates new variable per iteration
		funcs2[i] = func() {
			fmt.Print(i, " ")
		}
	}
	for _, fn := range funcs2 {
		fn()
	}
	fmt.Println()
}

// makeAdder returns a closure that adds x to its argument.
func makeAdder(x int) func(int) int {
	return func(y int) int {
		return x + y
	}
}

// counter_gen returns a closure that counts up from 1.
func counter_gen() func() int {
	n := 0
	return func() int {
		n++
		return n
	}
}
